File: CICADA_FIGURES/a3c_Fig_Swimmers_Obs_g_formula.m
clear all; clc; format compact

% FIGURE: G-FORMULA CORRECTED SURVIVAL CURVES (No swimmer plots)
figure(2); clf;

%% Load survival curves - computed by a1_CausalSurvivalAnalysis.m
load('bootstrap_confidence_bands.mat');

% Create axes using most of the figure space
left_margin = 0.15;
bottom_margin = 0.12;
width = 0.75;
height = 0.75;

ax = axes('Position', [left_margin, bottom_margin, width, height]);

% Plot observational survival curves with confidence intervals
hold on;

% Time vector (should be same for all curves)
t_grid = 0:2:168;  % This gives 85 points

% Convert to column vectors for consistency
t_grid = t_grid(:);
s0_lower = s0_lower(:);
s0_upper = s0_upper(:);
s1_lower = s1_lower(:);
s1_upper = s1_upper(:);

% Plot confidence bands first (so they appear behind the lines)
% Untreated confidence band (gray)
fill([t_grid; flipud(t_grid)], ...
     [s0_lower; flipud(s0_upper)], ...
     [0.7, 0.7, 0.7], 'FaceAlpha', 0.3, 'EdgeColor', 'none', ...
     'DisplayName', 'Untreated 95% CI');

% Treated confidence band (gray)
fill([t_grid; flipud(t_grid)], ...
     [s1_lower; flipud(s1_upper)], ...
     [0.7, 0.7, 0.7], 'FaceAlpha', 0.3, 'EdgeColor', 'none', ...
     'DisplayName', 'Treated 95% CI');

% Plot main curves
% Get reference curves from RCT data
T1 = readtable('trialData1.csv');
[s0_true, s1_true, t0_true, t1_true] = fcnPlotKM(T1);

% RCT curves for comparison - dashed lines with blue colors
t0 = 0:2:168; t1 = t0; 
plot(t0, s0_true, '--', 'Color', [0.4, 0.4, 0.8], 'LineWidth', 2.5);     % Medium blue dashed
plot(t1, s1_true, '--', 'Color', [0, 0, 0.5], 'LineWidth', 2.5);       % Dark blue dashed

% Observational estimated curves - solid lines with blue colors
plot(t0, s0_median, '-', 'Color', [0.4, 0.4, 0.8], 'LineWidth', 2.5);   % Medium blue solid
plot(t1, s1_median, '-', 'Color', [0, 0, 0.5], 'LineWidth', 2.5);     % Dark blue solid

% Add text labels using manually adjusted positions
text(148.22, 0.2435, 'Untreated', ...
    'Color', [0.4, 0.4, 0.8], 'FontSize', 12, 'FontWeight', 'bold', ...
    'HorizontalAlignment', 'center');
text(152.67, 0.5965, 'Treated', ...
    'Color', [0, 0, 0.5], 'FontSize', 12, 'FontWeight', 'bold', ...
    'HorizontalAlignment', 'center');

% Formatting
xlabel('Time (hours)', 'FontSize', 12);
ylabel('Survival Probability', 'FontSize', 12);
grid on;
set(gca, 'FontSize', 11);
xlim([0, 168]);
ylim([0, 1]);

% Set x-axis ticks every 24 hours
xticks(0:24:168);
xticklabels(0:24:168);
xlim([0 168]);

% Set figure size (in inches) - smaller since no swimmer plots
fig_width = 7;    % Width in inches
fig_height = 5;   % Height in inches (reduced from 9)

% Set figure properties
set(gcf, 'Units', 'inches');
set(gcf, 'Position', [1, 1, fig_width, fig_height]); % [left, bottom, width, height]
set(gcf, 'PaperUnits', 'inches');
set(gcf, 'PaperSize', [fig_width, fig_height]);
set(gcf, 'PaperPosition', [0, 0, fig_width, fig_height]);

% Export as PDF (vector format)
print(gcf, 'Fig_gformula_corrected_survival_curves.pdf', '-dpdf', '-r300');

%% Export G-formula corrected analysis results to text file for paper
filename = sprintf('gformula_corrected_analysis_results_%s.txt', datestr(now, 'yyyymmdd_HHMMSS'));
fid = fopen(filename, 'w');

fprintf(fid, '==========================================================\n');
fprintf(fid, 'G-FORMULA CORRECTED ANALYSIS RESULTS FOR PAPER\n');
fprintf(fid, 'Generated on: %s\n', datestr(now));
fprintf(fid, '==========================================================\n\n');

% Load RCT comparison data
T1 = readtable('trialData1.csv');
[s0_true, s1_true, t0_true, t1_true] = fcnPlotKM(T1);

% Study characteristics
fprintf(fid, 'STUDY CHARACTERISTICS:\n');
fprintf(fid, '- Analysis method: G-formula (parametric g-computation)\n');
fprintf(fid, '- Data source: Observational data with estimated parameters\n');
fprintf(fid, '- Uncertainty quantification: Bootstrap confidence intervals\n');
fprintf(fid, '- Comparison standard: RCT ground truth\n');
fprintf(fid, '- Study period: 168 hours\n\n');

% Primary outcomes comparison
gformula_untreated_survival = s0_median(end) * 100;
gformula_treated_survival = s1_median(end) * 100;
gformula_ate = (s1_median(end) - s0_median(end)) * 100;

rct_untreated_survival = s0_true(end) * 100;
rct_treated_survival = s1_true(end) * 100;
rct_ate = (s1_true(end) - s0_true(end)) * 100;

fprintf(fid, 'PRIMARY OUTCOMES COMPARISON (168 hours):\n');
fprintf(fid, '%-20s %-15s %-15s %-15s\n', 'Analysis Method', 'Untreated(%)', 'Treated(%)', 'ATE(%)');
fprintf(fid, '%-20s %-15s %-15s %-15s\n', repmat('-', 1, 20), repmat('-', 1, 15), repmat('-', 1, 15), repmat('-', 1, 15));
fprintf(fid, '%-20s %-15.1f %-15.1f %-15.1f\n', 'RCT (Ground Truth)', rct_untreated_survival, rct_treated_survival, rct_ate);
fprintf(fid, '%-20s %-15.1f %-15.1f %-15.1f\n', 'G-formula', gformula_untreated_survival, gformula_treated_survival, gformula_ate);

% Bias reduction assessment
bias_untreated = gformula_untreated_survival - rct_untreated_survival;
bias_treated = gformula_treated_survival - rct_treated_survival;
bias_ate = gformula_ate - rct_ate;

fprintf(fid, '\nBIAS ASSESSMENT:\n');
fprintf(fid, '- Untreated survival bias: %+.1f%% points\n', bias_untreated);
fprintf(fid, '- Treated survival bias: %+.1f%% points\n', bias_treated);
fprintf(fid, '- Treatment effect bias: %+.1f%% points\n', bias_ate);

% Relative bias
if rct_ate ~= 0
    relative_bias_ate = (bias_ate / rct_ate) * 100;
    fprintf(fid, '- Relative bias in ATE: %+.1f%%\n', relative_bias_ate);
end

% G-formula performance assessment
fprintf(fid, '\nG-FORMULA PERFORMANCE:\n');
if abs(bias_ate) < 2
    fprintf(fid, '✓ EXCELLENT bias correction (<2%% points)\n');
elseif abs(bias_ate) < 5
    fprintf(fid, '✓ GOOD bias correction (2-5%% points)\n');
elseif abs(bias_ate) < 10
    fprintf(fid, '⚠️  MODERATE bias correction (5-10%% points)\n');
else
    fprintf(fid, '✗ POOR bias correction (>10%% points)\n');
end

% Confidence interval analysis
gformula_untreated_ci_lower = s0_lower(end) * 100;
gformula_untreated_ci_upper = s0_upper(end) * 100;
gformula_treated_ci_lower = s1_lower(end) * 100;
gformula_treated_ci_upper = s1_upper(end) * 100;

fprintf(fid, '\nCONFIDENCE INTERVALS (95%% Bootstrap):\n');
fprintf(fid, 'Untreated survival:\n');
fprintf(fid, '  G-formula: %.1f%% [%.1f%%, %.1f%%]\n', ...
    gformula_untreated_survival, gformula_untreated_ci_lower, gformula_untreated_ci_upper);
fprintf(fid, '  RCT: %.1f%% (point estimate)\n', rct_untreated_survival);

fprintf(fid, 'Treated survival:\n');
fprintf(fid, '  G-formula: %.1f%% [%.1f%%, %.1f%%]\n', ...
    gformula_treated_survival, gformula_treated_ci_lower, gformula_treated_ci_upper);
fprintf(fid, '  RCT: %.1f%% (point estimate)\n', rct_treated_survival);

% Treatment effect with confidence interval
gformula_ate_ci_lower = (s1_lower(end) - s0_upper(end)) * 100;  % Conservative lower bound
gformula_ate_ci_upper = (s1_upper(end) - s0_lower(end)) * 100;  % Conservative upper bound

fprintf(fid, 'Treatment effect (ATE):\n');
fprintf(fid, '  G-formula: %.1f%% [%.1f%%, %.1f%%]\n', ...
    gformula_ate, gformula_ate_ci_lower, gformula_ate_ci_upper);
fprintf(fid, '  RCT: %.1f%% (point estimate)\n', rct_ate);

% Coverage assessment - does CI contain true value?
rct_covered_untreated = (rct_untreated_survival >= gformula_untreated_ci_lower) && ...
                       (rct_untreated_survival <= gformula_untreated_ci_upper);
rct_covered_treated = (rct_treated_survival >= gformula_treated_ci_lower) && ...
                     (rct_treated_survival <= gformula_treated_ci_upper);
rct_covered_ate = (rct_ate >= gformula_ate_ci_lower) && (rct_ate <= gformula_ate_ci_upper);

fprintf(fid, '\nCOVERAGE ASSESSMENT:\n');
if rct_covered_untreated
    fprintf(fid, '- Untreated survival CI covers RCT: YES\n');
else
    fprintf(fid, '- Untreated survival CI covers RCT: NO\n');
end
if rct_covered_treated
    fprintf(fid, '- Treated survival CI covers RCT: YES\n');
else
    fprintf(fid, '- Treated survival CI covers RCT: NO\n');
end
if rct_covered_ate
    fprintf(fid, '- Treatment effect CI covers RCT: YES\n');
else
    fprintf(fid, '- Treatment effect CI covers RCT: NO\n');
end

% Statistical significance
ate_significant = (gformula_ate_ci_lower > 0 && gformula_ate_ci_upper > 0) || ...
                  (gformula_ate_ci_lower < 0 && gformula_ate_ci_upper < 0);

fprintf(fid, '\nSTATISTICAL SIGNIFICANCE:\n');
if ate_significant
    fprintf(fid, '✓ Treatment effect is STATISTICALLY SIGNIFICANT\n');
    fprintf(fid, '  95%% CI excludes zero\n');
else
    fprintf(fid, '✗ Treatment effect is NOT statistically significant\n');
    fprintf(fid, '  95%% CI includes zero\n');
end

% Key time points analysis
key_times = [24, 48, 72, 96, 120, 144, 168];
fprintf(fid, '\nSURVIVAL COMPARISON AT KEY TIME POINTS:\n');
fprintf(fid, '%-8s %-12s %-12s %-12s %-12s %-12s\n', 'Time(h)', 'RCT Untrt', 'G-F Untrt', 'RCT Trt', 'G-F Trt', 'Bias ATE');
fprintf(fid, '%-8s %-12s %-12s %-12s %-12s %-12s\n', '-------', '----------', '----------', '--------', '--------', '--------');

for i = 1:length(key_times)
    time_point = key_times(i);
    
    % Find time indices
    time_idx = find(t_grid >= time_point, 1);
    [~, rct_idx0] = min(abs(t0_true - time_point));
    [~, rct_idx1] = min(abs(t1_true - time_point));
    
    if ~isempty(time_idx)
        gf_untrt_surv = s0_median(time_idx) * 100;
        gf_trt_surv = s1_median(time_idx) * 100;
        rct_untrt_surv = s0_true(rct_idx0) * 100;
        rct_trt_surv = s1_true(rct_idx1) * 100;
        
        rct_ate_timepoint = rct_trt_surv - rct_untrt_surv;
        gf_ate_timepoint = gf_trt_surv - gf_untrt_surv;
        bias_ate_timepoint = gf_ate_timepoint - rct_ate_timepoint;
        
        fprintf(fid, '%-8d %-12.1f %-12.1f %-12.1f %-12.1f %-12.1f\n', ...
            time_point, rct_untrt_surv, gf_untrt_surv, rct_trt_surv, gf_trt_surv, bias_ate_timepoint);
    end
end

% Methodology validation
fprintf(fid, '\nMETHODOLOGY VALIDATION:\n');
fprintf(fid, 'G-formula approach:\n');
fprintf(fid, '- Parametric g-computation using estimated models\n');
fprintf(fid, '- Parameter estimation from observational data\n');
fprintf(fid, '- RCT emulation with 50%% treatment probability\n');
fprintf(fid, '- Bootstrap resampling for uncertainty quantification\n');

% Model components assessment
fprintf(fid, '\nMODEL COMPONENTS:\n');
fprintf(fid, '- Disease progression model (L): Estimated from observational data\n');
fprintf(fid, '- PKPD model: Mixed-effects parameter estimation\n');
fprintf(fid, '- Mortality hazard model (Y): Logistic regression on observational data\n');
fprintf(fid, '- Censoring model (V): Set to zero for RCT emulation\n');

% Clinical interpretation
fprintf(fid, '\nCLINICAL INTERPRETATION:\n');

% Treatment recommendation based on G-formula results
if gformula_ate > 0 && ate_significant
    fprintf(fid, '✓ TREATMENT RECOMMENDED\n');
    fprintf(fid, '  - Statistically significant benefit observed\n');
    fprintf(fid, '  - G-formula corrects for confounding bias\n');
    nnt = 100 / gformula_ate;
    fprintf(fid, '  - Number needed to treat: %.1f patients\n', nnt);
elseif gformula_ate > 0 && ~ate_significant
    fprintf(fid, '⚠️  TREATMENT EFFECT UNCERTAIN\n');
    fprintf(fid, '  - Positive effect but not statistically significant\n');
    fprintf(fid, '  - Consider larger study or additional evidence\n');
elseif gformula_ate < 0 && ate_significant
    fprintf(fid, '✗ TREATMENT NOT RECOMMENDED\n');
    fprintf(fid, '  - Statistically significant harm observed\n');
    nnh = 100 / abs(gformula_ate);
    fprintf(fid, '  - Number needed to harm: %.1f patients\n', nnh);
else
    fprintf(fid, '○ NO CLEAR TREATMENT EFFECT\n');
    fprintf(fid, '  - Effect size small and not significant\n');
end

% Comparison with naive analysis (if available)
T0 = readtable('trialData0.csv');
if exist('trialData0.csv', 'file')
    fprintf(fid, '\nCOMPARISON WITH NAIVE ANALYSIS:\n');
    fprintf(fid, 'G-formula advantages:\n');
    fprintf(fid, '- Adjusts for confounding by age, SOFA, disease severity\n');
    fprintf(fid, '- Provides unbiased treatment effect estimates\n');
    fprintf(fid, '- Includes uncertainty quantification via bootstrap\n');
    fprintf(fid, '- Enables causal interpretation of results\n');
end

% Limitations and assumptions
fprintf(fid, '\nLIMITATIONS AND ASSUMPTIONS:\n');
fprintf(fid, '- Assumes no unmeasured confounding\n');
fprintf(fid, '- Relies on correct model specification\n');
fprintf(fid, '- Parameter estimation quality affects results\n');
fprintf(fid, '- Bootstrap assumes exchangeability of patients\n');
fprintf(fid, '- Simulation period limited to 168 hours\n');

% Model performance metrics
fprintf(fid, '\nMODEL PERFORMANCE METRICS:\n');

% Calculate overall agreement metrics
mean_abs_bias = mean(abs([bias_untreated, bias_treated, bias_ate]));
max_abs_bias = max(abs([bias_untreated, bias_treated, bias_ate]));

fprintf(fid, '- Mean absolute bias: %.1f%% points\n', mean_abs_bias);
fprintf(fid, '- Maximum absolute bias: %.1f%% points\n', max_abs_bias);
fprintf(fid, '- Coverage rate: %d/3 (%.0f%%) confidence intervals cover truth\n', ...
    sum([rct_covered_untreated, rct_covered_treated, rct_covered_ate]), ...
    mean([rct_covered_untreated, rct_covered_treated, rct_covered_ate])*100);

% Confidence interval widths
ci_width_untreated = gformula_untreated_ci_upper - gformula_untreated_ci_lower;
ci_width_treated = gformula_treated_ci_upper - gformula_treated_ci_lower;
ci_width_ate = gformula_ate_ci_upper - gformula_ate_ci_lower;

fprintf(fid, '\nUNCERTAINTY QUANTIFICATION:\n');
fprintf(fid, '- Untreated CI width: %.1f%% points\n', ci_width_untreated);
fprintf(fid, '- Treated CI width: %.1f%% points\n', ci_width_treated);
fprintf(fid, '- Treatment effect CI width: %.1f%% points\n', ci_width_ate);

% Precision assessment
if ci_width_ate < 10
    fprintf(fid, '- Precision: HIGH (CI width < 10%% points)\n');
elseif ci_width_ate < 20
    fprintf(fid, '- Precision: MODERATE (CI width 10-20%% points)\n');
else
    fprintf(fid, '- Precision: LOW (CI width > 20%% points)\n');
end

% Research implications
fprintf(fid, '\nRESEARCH IMPLICATIONS:\n');
fprintf(fid, '- Demonstrates effectiveness of G-formula for bias correction\n');
fprintf(fid, '- Validates causal inference approach in PKPD context\n');
fprintf(fid, '- Provides template for observational study analysis\n');
fprintf(fid, '- Shows importance of uncertainty quantification\n');

% Figure generation details
fprintf(fid, '\nFIGURE GENERATION:\n');
fprintf(fid, '- Figure type: G-formula survival curves with confidence bands\n');
fprintf(fid, '- Output file: Fig_gformula_corrected_survival_curves.pdf\n');
fprintf(fid, '- Format: PDF vector graphics (300 DPI)\n');
fprintf(fid, '- Dimensions: 7 × 5 inches\n');
fprintf(fid, '- Features: Bootstrap confidence bands, RCT comparison\n');
fprintf(fid, '- Colors: Blue tones with gray confidence regions\n');

% Technical specifications
fprintf(fid, '\nTECHNICAL SPECIFICATIONS:\n');
fprintf(fid, '- Analysis software: MATLAB\n');
fprintf(fid, '- Bootstrap data: bootstrap_confidence_bands.mat\n');
fprintf(fid, '- Time resolution: 2-hour intervals (85 time points)\n');
fprintf(fid, '- Confidence level: 95%% (2.5th and 97.5th percentiles)\n');
fprintf(fid, '- Survival curves: Kaplan-Meier estimator on simulated RCT\n');

% Data sources and processing
fprintf(fid, '\nDATA SOURCES AND PROCESSING:\n');
fprintf(fid, '- Parameter estimates: From observational data analysis\n');
fprintf(fid, '- Bootstrap samples: Patient-level resampling\n');
fprintf(fid, '- Simulation: RCT emulation using estimated parameters\n');
fprintf(fid, '- Comparison: Direct comparison with RCT ground truth\n');

fclose(fid);
fprintf('G-formula corrected analysis results exported to: %s\n', filename);

File: CICADA_FIGURES_PY/a3c_Fig_Swimmers_Obs_g_formula.py
"""a3c_Fig_Swimmers_Obs_g_formula.py

Port of CICADA_FIGURES/a3c_Fig_Swimmers_Obs_g_formula.m.

Produces Fig_gformula_corrected_survival_curves.pdf: g-formula corrected survival
curves with bootstrap 95% CIs, compared to RCT ground truth. No swimmer panels.
Uses bootstrap_confidence_bands.mat from the repo root.
"""

from __future__ import annotations

import os
import sys
import warnings
from datetime import datetime

import matplotlib
matplotlib.use("Agg")

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.io import loadmat

HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, ".."))
PY_HELPERS = os.path.join(REPO_ROOT, "python")
sys.path.insert(0, PY_HELPERS)

np.random.seed(0)

try:
    from fcnPlotKM import fcnPlotKM  # type: ignore
except Exception as exc:  # pragma: no cover
    print(f"[a3c] WARNING: could not import fcnPlotKM: {exc}")
    sys.exit(0)


def _load_bootstrap(path: str) -> dict | None:
    """Load bootstrap_confidence_bands.mat with h5py fallback."""
    try:
        d = loadmat(path, squeeze_me=True, struct_as_record=False)
        keys = ["s0_lower", "s0_median", "s0_upper",
                "s1_lower", "s1_median", "s1_upper", "t_grid"]
        return {k: np.asarray(d[k]).ravel() for k in keys if k in d}
    except NotImplementedError:
        pass
    except Exception as exc:
        print(f"[a3c] loadmat failed: {exc}; trying h5py fallback.")

    try:
        import h5py  # type: ignore

        out: dict = {}
        with h5py.File(path, "r") as f:
            for k in ["s0_lower", "s0_median", "s0_upper",
                      "s1_lower", "s1_median", "s1_upper", "t_grid"]:
                if k in f:
                    out[k] = np.asarray(f[k]).ravel()
        return out
    except Exception as exc:
        print(f"[a3c] h5py fallback failed: {exc}")
        return None


def main() -> int:
    boot_mat = os.path.join(REPO_ROOT, "bootstrap_confidence_bands.mat")
    trial1 = os.path.join(REPO_ROOT, "trialData1.csv")
    if not os.path.exists(boot_mat):
        print(f"[a3c] WARNING: missing {boot_mat}; skipping.")
        return 0
    if not os.path.exists(trial1):
        print(f"[a3c] WARNING: missing {trial1}; skipping.")
        return 0

    boot = _load_bootstrap(boot_mat)
    if not boot or not all(
        k in boot for k in [
            "s0_lower", "s0_median", "s0_upper",
            "s1_lower", "s1_median", "s1_upper",
        ]
    ):
        print(f"[a3c] WARNING: bootstrap arrays not found in {boot_mat}; skipping.")
        return 0

    s0_lower = boot["s0_lower"]
    s0_median = boot["s0_median"]
    s0_upper = boot["s0_upper"]
    s1_lower = boot["s1_lower"]
    s1_median = boot["s1_median"]
    s1_upper = boot["s1_upper"]

    # MATLAB uses t_grid = 0:2:168 (85 points) — reconstruct locally.
    t_grid = np.arange(0.0, 169.0, 2.0)
    if t_grid.size != s0_median.size:
        # Fall back to whatever was saved in the mat file.
        t_grid = np.asarray(boot.get("t_grid", t_grid)).ravel()

    # RCT ground truth
    T1 = pd.read_csv(trial1)
    s0_true, s1_true, t0_true, t1_true = fcnPlotKM(T1)

    fig = plt.figure(figsize=(7.0, 5.0))
    ax = fig.add_axes([0.15, 0.12, 0.75, 0.75])

    # Confidence bands (gray, behind)
    ax.fill_between(
        t_grid, s0_lower, s0_upper,
        color=(0.7, 0.7, 0.7), alpha=0.3, linewidth=0,
        label="Untreated 95% CI",
    )
    ax.fill_between(
        t_grid, s1_lower, s1_upper,
        color=(0.7, 0.7, 0.7), alpha=0.3, linewidth=0,
        label="Treated 95% CI",
    )

    # RCT reference (dashed)
    ax.plot(t0_true, s0_true, "--", color=(0.4, 0.4, 0.8), linewidth=2.5)
    ax.plot(t1_true, s1_true, "--", color=(0.0, 0.0, 0.5), linewidth=2.5)

    # G-formula medians (solid)
    ax.plot(t_grid, s0_median, "-", color=(0.4, 0.4, 0.8), linewidth=2.5)
    ax.plot(t_grid, s1_median, "-", color=(0.0, 0.0, 0.5), linewidth=2.5)

    # Manual label positions copied from MATLAB
    ax.text(148.22, 0.2435, "Untreated",
            color=(0.4, 0.4, 0.8), fontsize=12, fontweight="bold",
            ha="center")
    ax.text(152.67, 0.5965, "Treated",
            color=(0.0, 0.0, 0.5), fontsize=12, fontweight="bold",
            ha="center")

    ax.set_xlabel("Time (hours)", fontsize=12)
    ax.set_ylabel("Survival Probability", fontsize=12)
    ax.grid(True)
    ax.tick_params(labelsize=11)
    ax.set_xlim(0, 168)
    ax.set_ylim(0, 1)
    ax.set_xticks(np.arange(0, 169, 24))

    out_pdf = os.path.join(HERE, "Fig_gformula_corrected_survival_curves.pdf")
    fig.savefig(out_pdf, format="pdf", dpi=300)
    plt.close(fig)
    print(f"[a3c] Wrote {out_pdf}")

    # ---- text summary ---------------------------------------------------
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    txt = os.path.join(HERE, f"gformula_corrected_analysis_results_{stamp}.txt")
    try:
        gf_unt = float(s0_median[-1]) * 100.0
        gf_trt = float(s1_median[-1]) * 100.0
        gf_ate = gf_trt - gf_unt
        rct_unt = float(s0_true[-1]) * 100.0
        rct_trt = float(s1_true[-1]) * 100.0
        rct_ate = rct_trt - rct_unt
        gf_unt_lo = float(s0_lower[-1]) * 100.0
        gf_unt_hi = float(s0_upper[-1]) * 100.0
        gf_trt_lo = float(s1_lower[-1]) * 100.0
        gf_trt_hi = float(s1_upper[-1]) * 100.0
        ate_lo = (float(s1_lower[-1]) - float(s0_upper[-1])) * 100.0
        ate_hi = (float(s1_upper[-1]) - float(s0_lower[-1])) * 100.0

        with open(txt, "w") as f:
            f.write("=" * 58 + "\n")
            f.write("G-FORMULA CORRECTED ANALYSIS RESULTS FOR PAPER\n")
            f.write(f"Generated on: {datetime.now().isoformat(timespec='seconds')}\n")
            f.write("=" * 58 + "\n\n")
            f.write("STUDY CHARACTERISTICS:\n")
            f.write("- Analysis method: G-formula (parametric g-computation)\n")
            f.write("- Data source: Observational data with estimated parameters\n")
            f.write("- Uncertainty: Bootstrap 95% confidence intervals\n")
            f.write("- Comparison: RCT ground truth\n")
            f.write("- Study period: 168 hours\n\n")
            f.write("PRIMARY OUTCOMES (168 hours):\n")
            f.write(
                f"- RCT (Ground Truth): Untreated {rct_unt:.1f}%, "
                f"Treated {rct_trt:.1f}%, ATE {rct_ate:.1f}%\n"
            )
            f.write(
                f"- G-formula:          Untreated {gf_unt:.1f}% "
                f"[{gf_unt_lo:.1f}, {gf_unt_hi:.1f}], "
                f"Treated {gf_trt:.1f}% [{gf_trt_lo:.1f}, {gf_trt_hi:.1f}], "
                f"ATE {gf_ate:.1f}% [{ate_lo:.1f}, {ate_hi:.1f}]\n\n"
            )
            f.write("BIAS ASSESSMENT:\n")
            f.write(f"- Untreated survival bias: {gf_unt - rct_unt:+.1f}%\n")
            f.write(f"- Treated survival bias:   {gf_trt - rct_trt:+.1f}%\n")
            f.write(f"- ATE bias:                {gf_ate - rct_ate:+.1f}%\n\n")
            f.write("FIGURE GENERATION:\n")
            f.write("- Output file: Fig_gformula_corrected_survival_curves.pdf\n")
            f.write("- Format: PDF vector graphics (300 DPI)\n")
            f.write("- Dimensions: 7 x 5 inches\n")
        print(f"[a3c] Wrote {txt}")
    except Exception as exc:
        warnings.warn(f"[a3c] text summary failed: {exc}")

    return 0


if __name__ == "__main__":
    sys.exit(main())
